packages/realm/src/app-services/SyncSession.ts

```diff
@@ -229,106 +228,90 @@ const mockApp = new Proxy({} as App, {
   ListenerToken,
-  [binding.SyncSession, ProgressDirection, ProgressMode]
+  [binding.WeakSyncSession, binding.SyncSession, ProgressDirection, ProgressMode]
 >({
   throwOnReAdd: true,
-  add(callback, internal, direction, mode) {
+  add(callback, weakInternal, internal, direction, mode) {
     const token = internal.registerProgressNotifier(
       (transferred, transferable) => callback(Number(transferred), Number(transferable)),
       toBindingDirection(direction),
       mode === ProgressMode.ReportIndefinitely,
     );
-    return { internal, token };
+    return { weakInternal, token };
   },
-  remove({ internal, token }) {
-    return internal.unregisterProgressNotifier(token);
+  remove({ weakInternal, token }) {
+    weakInternal.withDeref((internal) => internal?.unregisterProgressNotifier(token));
   },
 });
 
 /**
  * Connection listeners are shared across instances of the SyncSession, making it possible to deregister a listener on another session
  * TODO: Consider adding a check to verify that the callback is removed from the correct SyncSession (although that would break the API)
  */
-const CONNECTION_LISTENERS = new Listeners<ConnectionNotificationCallback, ListenerToken, [binding.SyncSession]>({
+const CONNECTION_LISTENERS = new Listeners<
+  ConnectionNotificationCallback,
+  ListenerToken,
+  [binding.WeakSyncSession, binding.SyncSession]
+>({
   throwOnReAdd: true,
-  add(callback, internal) {
+  add(callback, weakInternal, internal) {
     const token = internal.registerConnectionChangeCallback((oldState, newState) =>
       callback(fromBindingConnectionState(newState), fromBindingConnectionState(oldState)),
     );
-    return { internal, token };
+    return { weakInternal, token };
   },
-  remove({ internal, token }) {
-    internal.unregisterConnectionChangeCallback(token);
+  remove({ weakInternal, token }) {
+    weakInternal.withDeref((internal) => internal?.unregisterConnectionChangeCallback(token));
   },
 });
 
 export class SyncSession {
-  private static instances = new Set<binding.WeakRef<SyncSession>>();
-
-  /**
-   * Resets the internal shared pointer of all instances returned since this message was last called.
-   * @internal
-   */
-  public static resetAllInternals() {
-    assert(flags.ALLOW_CLEAR_TEST_STATE, "Set the flags.ALLOW_CLEAR_TEST_STATE = true before calling this.");
-    for (const sessionRef of SyncSession.instances) {
-      sessionRef.deref()?.resetInternal();
-    }
-    SyncSession.instances.clear();
-  }
-
   /** @internal */
-  private _internal: binding.SyncSession | null;
+  private weakInternal: binding.WeakSyncSession;
   /** @internal */
-  public get internal() {
-    assert(this._internal, "This SyncSession is no longer valid");
-    return this._internal;
+  public withInternal<Ret = void>(cb: (syncSession: binding.SyncSession) => Ret) {
+    return this.weakInternal.withDeref((syncSession) => {
+      assert(syncSession, "This SyncSession is no longer valid");
```

Comment:
Note: this is the place that unconditionally throws from each method if the C++ object has been destroyed already. See the slack poll on #realm-js-team. This PR currently implements behavior (1) but I can easily change that if the consensus forms around a different option.

Reply:
Let's try this and see if it becomes too painful to work with 👍